### PR TITLE
Pagination: Add Close Cursor API in v2.

### PR DESCRIPTION
**Pull request by Yury-Fridlyand**
### Description
Add close cursor API to V2 similar to V1. Refer to [pagination doc](https://github.com/opensearch-project/sql/blob/main/docs/dev/opensearch-pagination.md#32-protocol).

This code branched of #1600. Once #1600 merged I'll update this branch and change base branch of this PR. Currently, it is a temporary branch to show the changes and approach.

### Design
https://github.com/opensearch-project/sql/blob/9100c6285f033784b5a2e58f53d575ece360737f/docs/dev/Pagination-v2.md#close-cursor

### Issues Resolved
New endpoint on `_plugins/_sql/close`. Under the hood it works as a subsequent query request, reserialized the cursor, but instead of executing it, just closes to encoded `scroll`.

### TODOs
- [x] Update docs
- [x] Add UT with coverage
- [x] Complete todos in code
- [ ] ~Return an error on repetitive `close` or on failure~
- [x] Don't fallback to V1 on `close` failure
- [ ] Record a demo
 
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass, including unit test, integration test and doctest
- [x] New functionality has been documented.
  - [ ] New functionality has javadoc added
  - [x] New functionality has user manual doc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: core/src/main/java/org/opensearch/sql/ast/tree/CloseCursor.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+import org.opensearch.sql.ast.AbstractNodeVisitor;
+import org.opensearch.sql.ast.Node;
+import org.opensearch.sql.executor.pagination.PlanSerializer;
+
+import java.util.List;
+
+/**
+ * TODO
```
Comment (acarbonetto) [inline core/src/main/java/org/opensearch/sql/ast/tree/CloseCursor.java:17]: TODO-bedo...bop!

File: core/src/main/java/org/opensearch/sql/ast/tree/CloseCursor.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+ */
+public class CloseCursor extends UnresolvedPlan {
+
+  /** An instance of {@link Cursor} */
```
Comment (acarbonetto) [inline core/src/main/java/org/opensearch/sql/ast/tree/CloseCursor.java:21]: This is a child or a cursor?  If it's a cursor, maybe call it a cursor...?

File: core/src/main/java/org/opensearch/sql/executor/execution/NonQueryPlan.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+import org.opensearch.sql.ast.tree.UnresolvedPlan;
+import org.opensearch.sql.common.response.ResponseListener;
+import org.opensearch.sql.executor.ExecutionEngine;
+import org.opensearch.sql.executor.QueryId;
+import org.opensearch.sql.executor.QueryService;
+
+/**
+ * Query plan which does not reflect a search query being executed.
+ */
+public class NonQueryPlan extends AbstractPlan {
```
Comment (acarbonetto) [inline core/src/main/java/org/opensearch/sql/executor/execution/NonQueryPlan.java:20]: Is there a better way to call this than "non-query"?  
Maybe "CursorPlan"...?  Or "EmptyPlan"?  
I'm not really sure...

File: core/src/main/java/org/opensearch/sql/executor/execution/QueryPlanFactory.java [hunk 4]
```diff
@@ -76,10 +77,15 @@ public AbstractPlan create(
   /**
    * Creates a ContinuePaginatedPlan from a cursor.
    */
-  public AbstractPlan create(String cursor, boolean isExplain,
+  public AbstractPlan create(String cursor, boolean isExplain, boolean isCursorClose,
```
Comment (acarbonetto) [inline core/src/main/java/org/opensearch/sql/executor/execution/QueryPlanFactory.java:80]: would it be smarter to have an action or resource type for this?  Instead of isExplain, or isCursorClose (etc) we could have action is an enum of type: 
- SEARCH
- EXPLAIN
- CURSOR_CLOSE
- ...etc

Comment (acarbonetto) [reply core/src/main/java/org/opensearch/sql/executor/execution/QueryPlanFactory.java:80]: Alternatively, use separate functions for each type of action: 
- `createSearch`
- `createExplain`
- `createCursorClose`
- ...etc

File: core/src/main/java/org/opensearch/sql/executor/execution/QueryPlanFactory.java [hunk 5]
```diff
@@ -76,10 +77,15 @@ public AbstractPlan create(
...
                              ResponseListener<ExecutionEngine.QueryResponse> queryResponseListener,
                              ResponseListener<ExecutionEngine.ExplainResponse> explainListener) {
     QueryId queryId = QueryId.queryId();
+    if (isCursorClose) {
+      return new NonQueryPlan(queryId, new CloseCursor().attach(new Cursor(cursor)),
+          queryService, queryResponseListener);
+      // TODO explain?
```
Comment (acarbonetto) [inline core/src/main/java/org/opensearch/sql/executor/execution/QueryPlanFactory.java:87]: Feels like isExplain and isCursorClose shouldn't be present at the same time...

File: core/src/main/java/org/opensearch/sql/planner/logical/LogicalCloseCursor.java [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+ */
+
+package org.opensearch.sql.planner.logical;
+
+import java.util.List;
+import lombok.EqualsAndHashCode;
+import lombok.ToString;
+
+/**
+ * TODO
```
Comment (acarbonetto) [inline core/src/main/java/org/opensearch/sql/planner/logical/LogicalCloseCursor.java:13]: TODO-fla-doo!

File: core/src/main/java/org/opensearch/sql/planner/physical/CursorCloseOperator.java [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+package org.opensearch.sql.planner.physical;
+
+import java.util.List;
+import lombok.RequiredArgsConstructor;
+import org.opensearch.sql.data.model.ExprValue;
+import org.opensearch.sql.executor.ExecutionEngine;
+import org.opensearch.sql.storage.TableScanOperator;
+
+// TODO add doc
```
Comment (acarbonetto) [inline core/src/main/java/org/opensearch/sql/planner/physical/CursorCloseOperator.java:14]: TODO add-doc!

File: integ-test/src/test/java/org/opensearch/sql/sql/PaginationIT.java [hunk 8]
```diff
@@ -76,4 +81,32 @@ public void testCursorTimeout() throws IOException, InterruptedException {
...
+
+  @Test
+  @SneakyThrows
+  public void testCloseCursor() {
+    var query = "SELECT * from " + TEST_INDEX_CALCS;
+    var response = new JSONObject(executeFetchQuery(query, 4, "jdbc"));
+    assertTrue(response.has("cursor"));
+    var cursor = response.getString("cursor");
+
+    Request closeCursorRequest = new Request("POST", "_plugins/_sql/close");
```
Comment (acarbonetto) [inline integ-test/src/test/java/org/opensearch/sql/sql/PaginationIT.java:93]: api resource should be `_close`?  Or is this following legacy API?

File: integ-test/src/test/java/org/opensearch/sql/sql/PaginationIT.java [hunk 9]
```diff
@@ -76,4 +81,32 @@ public void testCursorTimeout() throws IOException, InterruptedException {
 
     wipeAllClusterSettings();
   }
+
+  @Test
+  @SneakyThrows
+  public void testCloseCursor() {
+    var query = "SELECT * from " + TEST_INDEX_CALCS;
```
Comment (acarbonetto) [inline integ-test/src/test/java/org/opensearch/sql/sql/PaginationIT.java:88]: comment: "Test setup: Initial page request to get cursor"

File: integ-test/src/test/java/org/opensearch/sql/sql/PaginationIT.java [hunk 10]
```diff
@@ -76,4 +81,32 @@ public void testCursorTimeout() throws IOException, InterruptedException {
...
+    Request closeCursorRequest = new Request("POST", "_plugins/_sql/close");
+    closeCursorRequest.setJsonEntity(String.format("{ \"cursor\" : \"%s\" } ", cursor));
+    RequestOptions.Builder restOptionsBuilder = RequestOptions.DEFAULT.toBuilder();
+    restOptionsBuilder.addHeader("Content-Type", "application/json");
+    closeCursorRequest.setOptions(restOptionsBuilder);
+    response = new JSONObject(executeRequest(closeCursorRequest));
+    assertTrue(response.has("succeeded"));
+    assertTrue(response.getBoolean("succeeded"));
+
+    ResponseException exception =
```
Comment (acarbonetto) [inline integ-test/src/test/java/org/opensearch/sql/sql/PaginationIT.java:102]: add comment: 
"Test that cursor is no longer available"

File: opensearch/src/main/java/org/opensearch/sql/opensearch/request/OpenSearchScrollRequest.java [hunk 11]
```diff
@@ -165,7 +165,7 @@ public void writeTo(StreamOutput out) throws IOException {
 
       out.writeString(scrollId == null? "" : scrollId);
     }
-    out.writeBoolean(needClean);
+    //out.writeBoolean(needClean);
```
Comment (acarbonetto) [inline opensearch/src/main/java/org/opensearch/sql/opensearch/request/OpenSearchScrollRequest.java:168]: do you want to log this?

File: opensearch/src/main/java/org/opensearch/sql/opensearch/request/OpenSearchScrollRequest.java [hunk 12]
```diff
@@ -181,7 +181,7 @@ public OpenSearchScrollRequest(StreamInput in, OpenSearchStorageEngine engine)
     initialSearchRequest = new SearchRequest(in);
     scrollTimeout = in.readTimeValue();
     scrollId = in.readString();
-    needClean = in.readBoolean();
+    //needClean = in.readBoolean();
```
Comment (acarbonetto) [inline opensearch/src/main/java/org/opensearch/sql/opensearch/request/OpenSearchScrollRequest.java:184]: log this or remove?

File: protocol/src/main/java/org/opensearch/sql/protocol/response/format/NoQueryResponseFormatter.java [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.sql.protocol.response.format;
+
+import lombok.Getter;
+import org.opensearch.sql.protocol.response.QueryResult;
+
+public class NoQueryResponseFormatter extends JsonResponseFormatter<QueryResult> {
```
Comment (acarbonetto) [inline protocol/src/main/java/org/opensearch/sql/protocol/response/format/NoQueryResponseFormatter.java:11]: TODO: add javadoc

Comment (acarbonetto) [reply protocol/src/main/java/org/opensearch/sql/protocol/response/format/NoQueryResponseFormatter.java:11]: Would `EmptyResponseFormatter` be better?

File: sql/src/main/java/org/opensearch/sql/sql/domain/SQLQueryRequest.java [hunk 14]
```diff
@@ -112,6 +112,10 @@ public boolean isExplainRequest() {
     return path.endsWith("/_explain");
   }
 
+  public boolean isCursorCloseRequest() {
+    return path.endsWith("/close");
```
Comment (acarbonetto) [inline sql/src/main/java/org/opensearch/sql/sql/domain/SQLQueryRequest.java:116]: "/close" or "/_close"?

**Review by acarbonetto — COMMENTED** [review]
Thanks for this Yury

**Review by matthewryanwells — APPROVED** [review]

File: docs/dev/Pagination-v2.md [hunk 15]
```diff
@@ -669,6 +669,116 @@ PlanSerializer ->>+ CursorDeserializationStream : deserialize
...
+class CursorCloseOperator:
+  PhysicalPlan tree
+  def open:
+    pass
+    # no-op, don't propagate `open` of underlying plan tree
+
+  def hasNext:
+    return false
+```
+
```
Comment (Yury-Fridlyand) [inline docs/dev/Pagination-v2.md:781]: show default `open`

File: docs/dev/Pagination-v2.md [hunk 16]
```diff
@@ -669,6 +669,116 @@ PlanSerializer ->>+ CursorDeserializationStream : deserialize
...
+      state "..." as ...
+      state "OpenSearchIndexScan" as OpenSearchIndexScan
+
+      CursorCloseOperator --> ProjectOperator
+      ProjectOperator --> ...
+      ... --> OpenSearchIndexScan
+    }
+
+    [*] --> Unresolved : AstBuilder
+    Unresolved --> Abstract : QueryPlanner
```
Comment (Yury-Fridlyand) [inline docs/dev/Pagination-v2.md:749]: reorder unresolved and abstract

**Review by GumpacG — APPROVED** [review]

File: core/src/main/java/org/opensearch/sql/executor/execution/NonQueryPlan.java (hunk 3 above)
Comment (dai-chen) [reply core/src/main/java/org/opensearch/sql/executor/execution/NonQueryPlan.java:20]: yeah, I recall there are usually base class `Command` in other db for DDL or any statement that doesn't need planning. Otherwise many boilerplate code from AST to Physical will be added.

Comment (forestmvey) [reply core/src/main/java/org/opensearch/sql/executor/execution/NonQueryPlan.java:20]: "InertPlan"?

File: legacy/src/main/java/org/opensearch/sql/legacy/plugin/RestSQLQueryAction.java [hunk 17]
```diff
@@ -162,7 +163,11 @@ private ResponseListener<QueryResponse> createQueryResponseListener(
       BiConsumer<RestChannel, Exception> errorHandler) {
     Format format = request.format();
     ResponseFormatter<QueryResult> formatter;
-    if (format.equals(Format.CSV)) {
+
+    if (request.isCursorCloseRequest()) {
+      formatter = new NoQueryResponseFormatter();
+    }
```
Comment (forestmvey) [inline legacy/src/main/java/org/opensearch/sql/legacy/plugin/RestSQLQueryAction.java:169]: formatting.

**Review by forestmvey — APPROVED** [review]

File: integ-test/src/test/java/org/opensearch/sql/sql/PaginationIT.java (hunk 8 above)
Comment (Yury-Fridlyand) [reply integ-test/src/test/java/org/opensearch/sql/sql/PaginationIT.java:93]: Not a mistake:
https://github.com/opensearch-project/sql/blob/8e5d766b0fbea11b449eeada9fb1ccab375cec1f/legacy/src/main/java/org/opensearch/sql/legacy/plugin/RestSqlAction.java#L78-L80

File: sql/src/main/java/org/opensearch/sql/sql/domain/SQLQueryRequest.java (hunk 14 above)
Comment (Yury-Fridlyand) [reply sql/src/main/java/org/opensearch/sql/sql/domain/SQLQueryRequest.java:116]: https://github.com/opensearch-project/sql/pull/1660#discussion_r1207258495

File: core/src/main/java/org/opensearch/sql/executor/execution/QueryPlanFactory.java (hunk 4 above)
Comment (MaxKsyunz) [reply core/src/main/java/org/opensearch/sql/executor/execution/QueryPlanFactory.java:80]: We already have
- `QueryPlan`
- `ExplainPlan`
- time for `CommandPlan`, methinks.

File: sql/src/main/java/org/opensearch/sql/sql/SQLService.java [hunk 18]
```diff
@@ -73,7 +73,8 @@ private AbstractPlan plan(
             "is not supported. Use `explain` for the initial query request.");
       }
       return queryExecutionFactory.create(request.getCursor().get(),
-        isExplainRequest, queryListener.orElse(null), explainListener.orElse(null));
+        isExplainRequest, request.isCursorCloseRequest(),
```
Comment (MaxKsyunz) [inline sql/src/main/java/org/opensearch/sql/sql/SQLService.java:76]: [Echoing](https://github.com/opensearch-project/sql/pull/1660/files#r1206232124) Andrew's comments, this would look cleaner if `SQLService` analyzed a request and called appropriate method on `queryExecutionFactory` to create an `AbstractPlan`.

File: core/src/main/java/org/opensearch/sql/ast/tree/CloseCursor.java [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+  private UnresolvedPlan child;
```
Comment (MaxKsyunz) [inline core/src/main/java/org/opensearch/sql/ast/tree/CloseCursor.java:22]: What is this child used for? 

As far as I can see, this can be cursor id without loss of generality and would make the code more straight-forward.

File: protocol/src/main/java/org/opensearch/sql/protocol/response/format/NoQueryResponseFormatter.java (hunk 13 above)
Comment (MaxKsyunz) [reply protocol/src/main/java/org/opensearch/sql/protocol/response/format/NoQueryResponseFormatter.java:11]: `CommandResponseFormatter` to book-end `CommandPlan` abstract plan?

**Review by MaxKsyunz — COMMENTED** [review]
Thank you @Yury-Fridlyand!

What happens when closing the cursor fails?

File: core/src/main/java/org/opensearch/sql/ast/tree/CloseCursor.java (hunk 19 above)
Comment (Yury-Fridlyand) [reply core/src/main/java/org/opensearch/sql/ast/tree/CloseCursor.java:22]: True.
I tried to follow this graph:
image

If I store here the cursor itself instead of a child plan, the planner logic should be a bit more complicated to create 2 nodes.

**Review by MitchellGale — APPROVED** [review]

**Comment by Yury-Fridlyand** [comment]
> What happens when closing the cursor fails?

Nothing. Any exception is suppressed.

**Comment by GumpacG** [comment]
Seems that jacoco is failing

**Comment by Yury-Fridlyand** [comment]
Thanks, fixed.

File: core/src/main/java/org/opensearch/sql/ast/tree/CloseCursor.java (hunk 19 above)
Comment (MaxKsyunz) [reply core/src/main/java/org/opensearch/sql/ast/tree/CloseCursor.java]: I see how it works but it's hard to follow. I hope we can make it more explicit at some point later.

**Review by MaxKsyunz — APPROVED** [review]

File: core/src/main/java/org/opensearch/sql/executor/execution/CommandPlan.java [hunk 20]
```diff
@@ -0,0 +1,52 @@
...
+package org.opensearch.sql.executor.execution;
+
+import org.opensearch.sql.ast.tree.UnresolvedPlan;
+import org.opensearch.sql.common.response.ResponseListener;
+import org.opensearch.sql.executor.ExecutionEngine;
+import org.opensearch.sql.executor.QueryId;
+import org.opensearch.sql.executor.QueryService;
+
+/**
+ * Query plan which does not reflect a search query being executed.
```
Comment (acarbonetto) [inline core/src/main/java/org/opensearch/sql/executor/execution/CommandPlan.java:18]: We should describe this with what it _does_ have:  A QueryPlan with a command or action

**Review by acarbonetto — APPROVED** [review]

File: core/src/main/java/org/opensearch/sql/planner/physical/CursorCloseOperator.java [hunk 21]
```diff
@@ -0,0 +1,62 @@
...
+
+  /**
+   * Requires to provide a dummy schema, because this plan node located on the top of the tree.
+   */
+  @Override
+  public ExecutionEngine.Schema schema() {
+    return new ExecutionEngine.Schema(List.of());
+  }
+
+  // TODO remove
```
[the remark this answers is not in the document]
Comment (Yury-Fridlyand) [reply core/src/main/java/org/opensearch/sql/planner/physical/CursorCloseOperator.java:52]: #1649

**Review by acarbonetto — APPROVED** [review]

**Review by MaxKsyunz — APPROVED** [review]
